### PR TITLE
EKF: Miscellaneous yaw estimator improvements

### PR DESCRIPTION
**Pull request by priseborough**
Fixes the one frame delay in processing velocity measurements.
Updates EKF2_MAG_CAL param description to remove ambiguity when flying without compass use.
Eliminates unnecessary reset operations and state variables and uses more descriptive reset function name and description.
Reduces post takeoff EKF-GSF yaw drift due to large yaw gyro bias.
Fixes https://github.com/ArduPilot/ardupilot/issues/14654

Result from SITL test to verify yaw estimator operation with second reset induced by land, disarm, arm and takeoff again.

Yaw estimates:
Screen Shot 2020-04-25 at 10 19 17 am

Weights:
Screen Shot 2020-04-25 at 10 20 15 am

Velocity innovations:
Screen Shot 2020-04-25 at 10 21 17 am
Screen Shot 2020-04-25 at 10 21 55 am

Also

### CODE REVIEW

File: libraries/AP_NavEKF/EKFGSF_yaw.cpp [hunk 1]
```diff
@@ -176,16 +132,55 @@ void EKFGSF_yaw::update(const Vector3f &delAng,
         float yawDelta = wrap_PI(EKF[mdl_idx].X[2] - GSF.yaw);
         GSF.yaw_variance +=  GSF.weights[mdl_idx] * (EKF[mdl_idx].P[2][2] + sq(yawDelta));
     }
-
-    // prevent the same velocity data being used again
-    vel_data_updated = false;
 }
 
-void EKFGSF_yaw::pushVelData(Vector2f vel, float velAcc)
+void EKFGSF_yaw::fuseVelData(Vector2f vel, float velAcc)
```
Comment (rmackay9) [inline libraries/AP_NavEKF/EKFGSF_yaw.cpp:137]: it's not a big deal but you we can reduce some copying of memory by changing "Vector2f vel" to "const Vector2f &vel".  This changes it from "pass by value" to "pass by reference" but it is still "constant" so you don't need to worry about accidentally changing it within the function, that won't happen.

Comment (priseborough) [reply libraries/AP_NavEKF/EKFGSF_yaw.cpp]: Fixed

**Comment by priseborough** [comment]
There is work required to harmonise the AP_Compass and AP_Declination libraries to prevent a situation where they are returning inconsistent information if COMPASS_AUTODEC = 0, but I will tackle that in a separate PR. This was highlighted when trying to achieve flyaway testing by manually setting COMPASS_DEC

**Comment by priseborough** [comment]
A couple of quick flight tests using EKF3 and EK3_MAG_CAL = 7. 

Screen Shot 2020-04-25 at 1 58 23 pm

Screen Shot 2020-04-25 at 1 58 55 pm

This PR does not address the issue with getting preflight checks to pass when the magnetometer is disabled and EKF2 is being used.

**Comment by rmackay9** [comment]
we discussed briefly on the call but will review over the next week or so.

File: libraries/AP_NavEKF2/AP_NavEKF2.cpp [hunk 2]
```diff
@@ -245,7 +245,7 @@ const AP_Param::GroupInfo NavEKF2::var_info[] = {
...
...tes, when it will use a simpler magnetic heading fusion model that does not use magnetic field states and when it will use an alternative method of yaw determination to the magnetometer. The 3-axis magnetometer fusion is only suitable for use when the external magnetic field environment is stable. EK3_MAG_CAL = 0 uses heading fusion on ground, 3-axis fusion in-flight, and is the default setting for Plane users. EK3_MAG_CAL = 1 uses 3-axis fusion only when manoeuvring. EK3_MAG_CAL = 2 uses heading fusion at all times, is recommended if the external magnetic field is varying and is the default for rovers. EK3_MAG_CAL = 3 uses heading fusion on the ground and 3-axis fusion after the first in-air field and yaw reset has completed, and is the default for copters. EK3_MAG_CAL = 4 uses 3-axis fusion at all times. NOTE: The fusion mode can be forced to 2 for specific EKF cores using the EK2_MAG_MASK parameter. NOTE: limited operation without a magnetometer or any other yaw sensor is possible by setting all COMPASS_USE, COMPASS_USE2, COMPASS_USE3, etc parameters to 0 with COMPASS_ENABLE set to 1. If this is done, the EK2_GSF_RUN and EK2_GSF_USE masks must be set to the same as EK3_IMU_MASK.
```
Comment (tatsuy) [inline libraries/AP_NavEKF2/AP_NavEKF2.cpp:248]: I think that these parameter names are different.
EK3_MAG_CAL -> EK2_MAG_CAL
EK3_IMU_MASK -> EK2_IMU_MASK
EK2_GSF_RUN -> EK2_GSF_RUN_MASK
EK2_GSF_USE -> EK2_GSF_USE_MASK

**Comment by priseborough** [comment]
@rmackay9 ping

**Comment by priseborough** [comment]
I've added the fixes for https://github.com/ArduPilot/ardupilot/issues/14654

The following results are from SITL testing of a hand launch without use of the magnetomer.

EK3_ENABLE = 1
EK2_ENABLE = 0
AHRS_EKF_TYPE = 3
EK3_MAG_CAL = 7
COMPASS_ENABLE = 0
COMPASS_USE = 0
COMPASS_USE2 = 0
COMPASS_USE3 = 0

SITL was started using "sim_vehicle.py --frame plane-throw --console --map -L CMAC_PILOTSBOX" and launched using RC 7 2000 to trigger the throw after the "APM: EKF3 IMU0 origin set" message was received.

Here are the console messages and GPS innovations before the fix showing the delayed yaw alignment and larger than desirable velocity innovations:

2020-06-22 17:17:27 ArduPlane V4.1.0dev (07ad4798)
2020-06-22 17:17:27 842f139c1c526d45bafd9f12e8b4c5f0
2020-06-22 17:17:27 Param space used: 581/3840
2020-06-22 17:17:27 RC Protocol: SITL
2020-06-22 17:17:27 New mission
2020-06-22 17:17:27 New rally
2020-06-22 17:17:27 GPS 1: detected as u-blox at 230400 baud
2020-06-22 17:17:32 Armed AUTO, xaccel = 19.4 m/s/s, waiting 0.2 sec
2020-06-22 17:17:33 Triggered AUTO. GPS speed = 0.0
2020-06-22 17:17:33 Takeoff to 30m at 100.0m to 1.0 deg
2020-06-22 17:17:42 EKF3 IMU0 yaw aligned using GPS
2020-06-22 17:17:42 EKF3 IMU0 is using GPS

Screen Shot 2020-06-22 at 6 37 53 pm

After the fixes, the reset is a second after launch and the innovations are small  due to the more accurate yaw:

2020-06-22 18:37:04 ArduPlane V4.1.0dev (e64d786b)
2020-06-22 18:37:04 842f139c1c526d45bafd9f12e8b4c5f0
2020-06-22 18:37:04 Param space used: 581/3840
2020-06-22 18:37:04 RC Protocol: SITL
2020-06-22 18:37:04 New mission
2020-06-22 18:37:04 New rally
2020-06-22 18:37:04 GPS 1: detected as u-blox at 230400 baud
2020-06-22 18:37:06 Armed AUTO, xaccel = 19.4 m/s/s, waiting 0.2 sec
2020-06-22 18:37:07 Triggered AUTO. GPS speed = 0.0
2020-06-22 18:37:07 Takeoff to 30m at 100.0m to 0.2 deg
2020-06-22 18:37:08 EKF3 IMU0 yaw aligned using GPS
2020-06-22 18:37:08 EKF3 IMU0 is using GPS

Screen Shot 2020-06-22 at 6 41 50 pm

File: libraries/AP_NavEKF2/AP_NavEKF2.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_NavEKF2/AP_NavEKF2.cpp:248]: you're right, paul?

File: libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp [hunk 3]
```diff
@@ -264,7 +264,9 @@ void NavEKF3_core::SelectMagFusion()
     if ((effectiveMagCal == MagCal::GSF_YAW || !use_compass()) &&
         effectiveMagCal != MagCal::EXTERNAL_YAW &&
         effectiveMagCal != MagCal::EXTERNAL_YAW_FALLBACK) {
-        if (!yawAlignComplete) {
+
+        // because this type of reset event is not as time critical, require a continuous history of valid estimates
+        if (!yawAlignComplete && EKFGSF_yaw_valid_count >= GSF_YAW_VALID_HISTORY_THRESHOLD) {
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp:269]: EKFGSF_yaw_valid_count is only incremented when gpsDataToFuse is true. Do we want to support the GSF for vision based position systems?

Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp:269]: If they have velocity data, then we could, but  given the estimated yaw yaw would be relative to vision earth frame and the vision systems already provide a yaw of body wrt that frame fo reference, there doesn't seem to be much use for it at the moment. I would prefer to tackle that as separate PR with test support from the vision user community.

**Comment by vierfuffzig** [comment]
@priseborough @tridge i‘m following this closely as i do fly my trad. FW airframes without mag pretty exclusively. i‘ve done a lot of flights on EK3 now and did notice that after launching it takes a couple of seconds for EK3 yaw to align (2-4 secs usually), as compared to just a few ms on EK2. however, in real-flight conditions (hand launch) i can‘t seem to find the same large innovations and delay that SITL produces. 

on a sidenote i wonder if there‘s a more intuitive way to handle yaw-source on no-mag setups using EK3. there‘s been no separate setting required on EK2 too besides unchecking MAGn_USE. the current EK3 handling imho might be prone to cause relevant support effort.

**Comment by priseborough** [comment]
> @priseborough @tridge i‘m following this closely as i do fly my trad. FW airframes without mag pretty exclusively. i‘ve done a lot of flights on EK3 now and did notice that after launching it takes a couple of seconds for EK3 yaw to align (2-4 secs usually), as compared to just a few ms on EK2. however, in real-flight conditions (hand launch) i can‘t seem to find the same large innovations and delay that SITL produces:

This PR should fix the delayed yaw alignment with EKF3.

**Comment by priseborough** [comment]
@vierfuffzig I can modify the EKF3 params so that the EKF-GSF yaw use is enabled automatically if compass is disabled. I think you are right that having to set it to a specific value will increase support queries.

**Comment by vierfuffzig** [comment]
@priseborough thanks, i think moving GSF to auto will add a lot of convenience.

**Comment by priseborough** [comment]
I've added patch to make the use of the GPS for yaw automatic if COMPASS_USEx and COMPASS_ENABLE are set to 0. This has been tested in SITL using

../Tools/autotest/sim_vehicle.py --console --map --frame plane-throw -L CMAC_PILOTSBOX

to perform a hand launch to the west. EK3_MAG_CAL was set to 0 as was COMPASS_USEx and COMPASS_ENABLE.

The Copter use case was also tested in SITL with COMPASS_USEx and COMPASS_ENABLE set to 0. A takeoff in ALT_HOLD was performed, followed by horizontal movement to align the yaw angle.

@vierfuffzig If you are able to do testing with your setup and share the results, then that would be appreciated.

**Comment by vierfuffzig** [comment]
@priseborough thanks a lot, that's great! i might get the chance to get some flights in later today, will report back.

**Comment by vierfuffzig** [comment]
@priseborough are you 100% positive that COMPASS_ENABLE needs to be set to 1 on EK2 for no-mag use?
`setting all COMPASS_USE, COMPASS_USE2, COMPASS_USE3, etc parameters to 0 with COMPASS_ENABLE set to 1`
as it's 0 on EK3 re MAG_CAL param description.

**Comment by vierfuffzig** [comment]
@priseborough tested this on EK3 with 
MAG_CAL = 0
COMPASS_ENABLE = 0
COMPASS_USE = 0
COMPASS_USE2 = 0
COMPASS_USE3 = 0

image

fast yaw alignment within a few ms into flight, MAG_CAL on default works. excellent, thanks a lot!

**Comment by priseborough** [comment]
I did some testing with my Nano Goblin today running a MatekF765 wing board and on one launch ran into a condition where the yaw aligned rapidly, but the use of GPS was delayed a few seconds due to the bias state convergence check. Commit https://github.com/ArduPilot/ardupilot/pull/14191/commits/4799fbad5a5284b633eed24272f5fa40b4d2bc14 fixes this.

**Comment by vierfuffzig** [comment]
@priseborough did a couple of launches on my nanoTalon yesterday, will take a closer look at each log, don‘t think i ran into delayed GPS fusion though from what i saw on OSD.

plus i think i missed a previous PR of yours that already fixed the requirement to set MAG_CAL = 7 explicitely. sorry bout that.

**Comment by priseborough** [comment]
I've rebased and fixed the conflict in the AP_NavEKF2.cpp EK2_MAG_CAL parameter description

**Comment by peterbarker** [comment]
Well, it passed CI.

**Pull request merged**